lesson19.md

```diff
@@ -0,0 +1,320 @@
+<!-- .slide: id="lesson19" -->
+
+# Basic Frontend - Spring 2020
+
+Lesson 19, Tuesday, 2020-11-03
+
+---
+
+### Recap - Data types
+
+```js
+let a1 = 42;
+let a2 = true;
+let a3 = "hello";
+let a4 = function() {};
+let a5;
+let a6 = [];
+let a7 = {};
+```
+
+---
+
+### Solution - Data types
+
+```js
+let a1 = 42;       // number
+let a2 = true;     // boolean
+let a3 = "hello";  // string
+let a4 = function() {};  // function
+let a5;            // undefined
+let a6 = [];       // object (array)
+let a7 = {};       // object
+```
+
+---
+
+### Null
+
+* There's one datatype that we haven't learned about yet: `null`
+* `null` is like `undefined`. It indicates the absence of a value.
+* `null` can be used to indicate an *intentional* absence of an object value.
+* Example: `document.getElementById("NoExist");` returns `null`
+
+---
+
+### Recap - DOM
+
+* DOM is an API
+* It allows us to add/remove/change any element of a web page
+* We can make TODO lists and cool web shops
+* But how would we create a web page with "dynamic" data, for example today's weather?
+
+---
+
+<!-- .slide: id="lesson19:WebAPI" -->
+
+# Web APIs
+
+---
+
+### Basic building blocks
+
+We need to first learn about a few concepts:
+
+1. JSON - How to exchange data
+1. REST API - How to talk with a server
+1. `fetch` API - How to connect to a server
+
+---
+
+### Object notation
+
+* In JavaScript, we can put all keys of objects in quotes. This is purely optional:
+
+```js
+let me = {
+    "name": "John",
+    lastName: "Doe",
+    "hobbies": [ "Eat", "Sleep" ]
+};
+
+console.log(me.name); // "John"
+```
+
+---
+
+### JSON
+
+* JSON stands for **J**ava**S**cript **O**bject **N**otation
+* It is a string representation of a JavaScript object
+* All *keys* in our key value pairs must be in quotes
+* Must contain only these types:
+   * `string`, `number`, `boolean`, `array`
+   * `null`
+   * another JSON `object`
+* These types are **forbidden**:
+   * `function`, `undefined`
+
+---
+
+### Why JSON?
+
+* When we exchange data between a browser and a server, we can only use strings (not objects, numbers, booleans...)
```

Comment:
i'm pretty sure i can exchange binary data, as well

Reply:
strings can contain binary data. At least that's what we had to use before those fancy arrayBuffers :)

Reply:
mmh. can't I send a file via js as binary data? or do I have to decode it as base64 or something?

But then yes, nowadays there is probably some implementation to send protobuf messages from the browser, right?

Reply:
It's complicated :)

https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Sending_and_Receiving_Binary_Data#Receiving_binary_data_in_older_browsers

Reply:
:sob:

Reply:
I've reworded it a bit, hope it's less ambiguous :)